src/Preheat.cpp

```diff
@@ -0,0 +1,160 @@
+    }
+    return std::max(0.0, time);
+}
+
+double Preheat::getTemp(unsigned int extruder, double flow, bool is_initial_layer)
+{
+    if (is_initial_layer)
+    {
+        return config_per_extruder[extruder].material_print_temperature_layer_0;
+    }
+    return config_per_extruder[extruder].flow_temp_graph.getTemp(flow, config_per_extruder[extruder].material_print_temperature, config_per_extruder[extruder].flow_dependent_temperature);
+}
+
+Preheat::WarmUpResult Preheat::timeBeforeEndToInsertPreheatCommand_coolDownWarmUp(double time_window, unsigned int extruder, double temp0, double temp1, double temp2, bool during_printing)
+{
+    WarmUpResult result;
+    const Config& config = config_per_extruder[extruder];
+    double time_to_cooldown_1_degree = config.time_to_cooldown_1_degree[during_printing];
+    double time_to_heatup_1_degree = config.time_to_heatup_1_degree[during_printing];
+    result.total_time_window = time_window;
+
+    //                  ,temp2
+    //                 /                                    .
+    //     ,temp0     /                                     .
+    //      \ ' ' ' '/ ' ' '> outer_temp                    .
+    //       \______/                                       .
+    //               "-> temp1
+    //      ^^^^^^^^^^
+    //      limited_time_window
+    double outer_temp;
+    double limited_time_window;
+    if (temp0 < temp2)
+    { // extra time needed during heating
+        double extra_heatup_time = (temp2 - temp0) * time_to_heatup_1_degree;
+        result.heating_time = extra_heatup_time;
+        limited_time_window = time_window - extra_heatup_time;
+        outer_temp = temp0;
+    }
+    else
+    {
+        double extra_cooldown_time = (temp0 - temp2) * time_to_cooldown_1_degree;
+        result.heating_time = 0;
+        limited_time_window = time_window - extra_cooldown_time;
+        outer_temp = temp2;
+    }
+    double time_ratio_cooldown_heatup = time_to_cooldown_1_degree / time_to_heatup_1_degree;
+    double time_to_heat_from_standby_to_print_temp = getTimeToGoFromTempToTemp(extruder, temp1, outer_temp, during_printing);
+    double time_needed_to_reach_standby_temp = time_to_heat_from_standby_to_print_temp * (1.0 + time_ratio_cooldown_heatup);
+    if (time_needed_to_reach_standby_temp < limited_time_window)
+    {
+        result.heating_time += time_to_heat_from_standby_to_print_temp;
+        result.lowest_temperature = temp1;
+    }
+    else 
+    {
+        result.heating_time += limited_time_window * time_to_heatup_1_degree / (time_to_cooldown_1_degree + time_to_heatup_1_degree);
+        result.lowest_temperature = std::max(temp1, temp2 - result.heating_time / time_to_heatup_1_degree);
+    }
+    return result;
+}
+
+Preheat::CoolDownResult Preheat::timeBeforeEndToInsertPreheatCommand_warmUpCoolDown(double time_window, unsigned int extruder, double temp0, double temp1, double temp2, bool during_printing)
```

Comment:
`warmUpCoolDown` doesn't describe anything in the context of regulating the temperature.